File: leetcode-04/solutions/cpp/0039.cpp
class Solution {
 public:
  vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
    vector<vector<int>> ans;
    vector<int> path;

    sort(candidates.begin(), candidates.end());
    dfs(candidates, target, 0, path, ans);

    return ans;
  }

 private:
  void dfs(vector<int>& candidates, int target, int s, vector<int>& path,
           vector<vector<int>>& ans) {
    if (target < 0) return;
    if (target == 0) {
      ans.push_back(path);
      return;
    }

    for (int i = s; i < candidates.size(); i++) {
      path.push_back(candidates[i]);
      dfs(candidates, target - candidates[i], i, path, ans);
      path.pop_back();
    }
  }
};

File: leetcode-04/solutions/python3/0039.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []

        candidates.sort()
        self.dfs(candidates, target, 0, [], ans)

        return ans

    def dfs(self, candidates: List[int], target: int, s: int, path: List[int], ans: List[List[int]]) -> None:
        if target < 0:
            return
        if target == 0:
            ans.append(path)
            return

        for i in range(s, len(candidates)):
            self.dfs(candidates, target - candidates[i], i, path + [candidates[i]], ans)
